### DISCUSSION TITLE
Drag & Drop / Window Focus

### DISCUSSION THREAD

**Discussion by LMSSonos** [tui-cs/Terminal.Gui#3314] [Q&A] [answered]
This question is more c# console specific than probably Terminal.Gui. 

I have implemented drag & drop with a `TextField` and attached the `TextChanging` event. In that event I call `File.Exists()` for the `TextChangingEventArgs.NewText` once it is true, i handle it as if it was a drop. This works very well.

My main issue is, that in that case i set the focus to another TextField where I can enter the description, the thing I'm struggling is to bring the window to front and activate it, so that I just can go on typing.  I tried the `BringWindowToTop` which works to bring the window on top, but I did not find any way to activate it.

**Comment by BDisp** [comment]
After bring the window on top call `<yourTextField>.SetFocus()`.

**Comment by LMSSonos** [comment]
Thanks for your answer. I already have the .SetFocus() but I was looking for something like following snipped (works in Windows, but not Linux)

```
var proc = Process.GetCurrentProcess();
SetForegroundWindow(proc.MainWindowHandle);
```

**Reply by BDisp** [reply, marked as answer]
That is low level code winuser.h which really only works on Windows. For Linux you've to look an analog library, probably using the libX11.so. I really don't know if the code below works, but could give you an hint.

```cs
using System;
using X11;

class FocusX11
{
    static void Main()
    {
        IntPtr display = Xlib.XOpenDisplay(null);
        if (display == IntPtr.Zero) return;

        // Obtain your window ID somehow
        IntPtr window = new IntPtr(0x123456); 

        Xlib.XRaiseWindow(display, window);
        Xlib.XSetInputFocus(display, window, (int)Xlib.RevertToPointerRoot, IntPtr.Zero);
        Xlib.XFlush(display);
    }
}
```

**Reply by LMSSonos** [reply]
Thank you very much for your suggestion and time. 

I will mark this as answer, because I did some research as well and as I'm on wayland it's again different way to handle it. As far as I could figure out, there is no easy way to get the window id in wayland with c# in a console app.

**Reply by BDisp** [reply]
Wayland deliberately does not allow arbitrary focus stealing for security/usability. Window focus is controlled by the compositor.
